server/src/main/java/org/opensearch/wlm/QueryGroupService.java

```diff
@@ -205,16 +214,53 @@ public void incrementFailuresFor(final String queryGroupId) {
         final Map<String, QueryGroupStatsHolder> statsHolderMap = new HashMap<>();
-        for (Map.Entry<String, QueryGroupState> queryGroupsState : queryGroupsStateAccessor.getQueryGroupStateMap().entrySet()) {
-            final String queryGroupId = queryGroupsState.getKey();
-            final QueryGroupState currentState = queryGroupsState.getValue();
-
-            statsHolderMap.put(queryGroupId, QueryGroupStatsHolder.from(currentState));
+        Map<String, QueryGroup> existingGroups = clusterService.state().metadata().queryGroups();
+        if (!queryGroupIds.contains("_all")) {
+            for (String id : queryGroupIds) {
+                if (!existingGroups.containsKey(id)) {
+                    throw new ResourceNotFoundException("QueryGroup with id " + id + " does not exist");
+                }
+            }
         }
+        Map<String, QueryGroupState> stateMap = queryGroupsStateAccessor.getQueryGroupStateMap();
+        if (stateMap != null) {
+            stateMap.forEach((queryGroupId, currentState) -> {
+                boolean shouldInclude = queryGroupIds.contains("_all") || queryGroupIds.contains(queryGroupId);
+                if (shouldInclude) {
+                    if (requestedBreached == null
+                        || requestedBreached == (resourceLimitBreached(existingGroups.get(queryGroupId), currentState).v1()
+                            .length() != 0)) {
+                        statsHolderMap.put(queryGroupId, QueryGroupStatsHolder.from(currentState));
+                    }
+                }
+            });
+        }
+        return new WlmStats(transportService.getLocalNode(), new QueryGroupStats(statsHolderMap));
+    }
 
-        return new QueryGroupStats(statsHolderMap);
+    /**
+     * @return if the QueryGroup breaches any resource limit based on the LastRecordedUsage
+     */
+    public Tuple<StringBuilder, ResourceType> resourceLimitBreached(QueryGroup queryGroup, QueryGroupState queryGroupState) {
+        StringBuilder reason = new StringBuilder();
+        for (ResourceType resourceType : TRACKED_RESOURCES) {
+            if (queryGroup.getResourceLimits().containsKey(resourceType)) {
+                final double threshold = getNormalisedRejectionThreshold(queryGroup.getResourceLimits().get(resourceType), resourceType);
+                final double lastRecordedUsage = queryGroupState.getResourceState().get(resourceType).getLastRecordedUsage();
+                if (threshold < lastRecordedUsage) {
+                    reason.append(resourceType)
+                        .append(" limit is breaching for ENFORCED type QueryGroup: (")
+                        .append(threshold)
+                        .append(" < ")
+                        .append(lastRecordedUsage)
+                        .append("). ");
+                    return new Tuple<>(reason, resourceType);
+                }
+            }
+        }
+        return new Tuple<>(reason, null);
```

Comment:
Instead of reusing this method and create unnecessary reason objects in memory, we should write separate one that just returns the boolean for efficiency